client/src/components/WebAuthn/Register.js

```diff
@@ -36,17 +36,49 @@ class Register extends Component {
     this.handleStartRegistration = this.handleStartRegistration.bind(this);
   }
 
+/**
+ * We rely on a prop called `keyData` being present, however it may not be set upon the first
+ * render of the component. We can check and wait for it via state.
+ *
+ * React 16 deprecates `componentWillReceiveProps` replacing it with `getDerivedStateFromProps`
+ * for setting state in a functional manner. We have chosen to use {@see componentDidUpdate} over
+ * `componentWillReceiveProps` in this component in order to provide the same functionality across
+ * different (supported) build versions.
+ * By having one delegate to the other we still save in maintenance costs.
+ *
+ * @param {Object} newProps incoming props for this component
+ * @param {Object} currentState current stated of this component
+ */
+  static getDerivedStateFromProps(newProps, currentState) {
+    if (currentState.view === VIEWS.LOADING && newProps.keyData) {
+      return { view: VIEWS.READY };
+    }
+    return null;
+  }
+
   /**
-   * keyData may not be passed through on initial render, and we can't start without it.
+   * React docs (as at React 16) imply that one should try to avoid setting state in this method...
+   * however if one is to do this, then it must be wrapped in a conditional in order to avoid
+   * an infinite loop. This method has been chosen insead of `componentWillReceieveProps` which
+   * is arguably the correct way to approach this, but it is deprecated in React 16 - we need
+   * this component to function in both React 16 and React 15 (used by the CMS still @ 4.4.1).
+   *
+   * Quote from https://reactjs.org/docs/react-component.html#componentdidupdate :
    *
-   * @todo Re-evaluate whether this code is required.
+   * > You **may call `setState()` immediately** in componentDidUpdate() but note that
+   * > **it must be wrapped in a condition** like in the example above, or you’ll cause an
+   * > infinite loop. It would also cause an extra re-rendering which, while not visible to the
+   * > user, can affect the component performance. If you’re trying to “mirror” some state to a
+   * > prop coming from above, consider using the prop directly instead.
+   * > [Read more about why copying props into state causes bugs.][1]
+   * >
+   * > [1]: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
    */
-  componentWillReceiveProps() {
-    const { keyData } = this.props;
-    const { view } = this.state;
-
-    if (view === VIEWS.LOADING && keyData) {
-      this.setState({ view: VIEWS.READY });
+  componentDidUpdate() {
+    const stateChange = Register.getDerivedStateFromProps(this.props, this.state);
```

Comment:
I don't think this needs to be in its own method TBH. How about:

```jsx
componentDidUpdate() {
  if (this.state.view === VIEWS.LOADING && this.props.keyData) {
    /* eslint-disable-next-line react/no-did-update-set-state */
    this.setState({ view: VIEWS.READY });
  }
}
```

Reply:
The second method was for react 16 compatibility.
However it's removed now.